kernel/mmanutils.c

```diff
@@ -194,106 +194,127 @@ static ssize_t _map_file_onto_memory(
-            return (void*)-1;
+            ERAISE(-EACCES);
+
+        /* fail is offset is not page-aligned */
+        if (offset % PAGE_SIZE)
+            ERAISE(-EINVAL);
     }
 
     if (fd >= 0 && addr)
     {
         ssize_t n;
+
         // ATTN: call mmap or mremap here so that this range refers to
         // a mapped region.
-        // ATTN: is non-page-aligned length valid?
+        // ATTN: Use EACCES for failures related to fd access. This may not
+        // conform the Linux kernel behavior.
         if (myst_mman_mprotect(&_mman, addr, length, prot | MYST_PROT_WRITE))
-            return (void*)-1;
+            ERAISE(-EACCES);
         if ((n = _map_file_onto_memory(fd, offset, addr, length, flags)) < 0)
-            return (void*)-1;
+            ERAISE(-EACCES);
         if (!(prot & MYST_PROT_WRITE))
         {
             if (myst_mman_mprotect(&_mman, addr, length, prot))
-                return (void*)-1;
+                ERAISE(-EACCES);
         }
-        void* end = (uint8_t*)addr + length;
-        assert(addr >= _mman_start && addr <= _mman_end);
-        assert(end >= _mman_start && end <= _mman_end);
 
-        return addr;
+        ret = (long)addr;
     }
-
-    int tflags = 0;
-
-    if (flags & MYST_MAP_FIXED)
-        tflags = MYST_MAP_ANONYMOUS | MYST_MAP_PRIVATE | MYST_MAP_FIXED;
     else
-        tflags = MYST_MAP_ANONYMOUS | MYST_MAP_PRIVATE;
+    {
+        int tflags = 0;
 
-    if ((r = myst_mman_mmap(&_mman, addr, length, prot, tflags, &ptr)) < 0)
-        return (void*)(long)r;
+        if (flags & MYST_MAP_FIXED)
+            tflags = MYST_MAP_ANONYMOUS | MYST_MAP_PRIVATE | MYST_MAP_FIXED;
+        else
+            tflags = MYST_MAP_ANONYMOUS | MYST_MAP_PRIVATE;
 
-    if (fd >= 0 && !addr)
-    {
-        ssize_t n;
-        // ATTN: is non-page-aligned length valid?
-        if (!(prot & MYST_PROT_WRITE))
-        {
-            if (myst_mman_mprotect(&_mman, ptr, length, prot | MYST_PROT_WRITE))
-                return (void*)-1;
-        }
-        if ((n = _map_file_onto_memory(fd, offset, ptr, length, flags)) < 0)
-            return (void*)(long)-n;
-        if (!(prot & MYST_PROT_WRITE))
+        ECHECK(
+            myst_mman_mmap(&_mman, addr, length, prot, tflags, (void**)&ret));
+
+        if (fd >= 0 && !addr)
         {
-            if (myst_mman_mprotect(&_mman, ptr, length, prot))
-                return (void*)-1;
+            ssize_t n;
+
+            // ATTN: Use EACCES for failures related to fd access. This may not
+            // conform the Linux kernel behavior.
+            if (!(prot & MYST_PROT_WRITE))
+            {
+                if (myst_mman_mprotect(
+                        &_mman, (void*)ret, length, prot | MYST_PROT_WRITE))
+                    ERAISE(-EACCES);
+            }
+            if ((n = _map_file_onto_memory(
+                     fd, offset, (void*)ret, length, flags)) < 0)
+                ERAISE((unsigned long)-n);
```

Comment:
`n` is negative, shouldn't it be `ERAISE(n)`?

Reply:
updated.